File: crates/compiler/frontend/python/apxm_program/_bridge.py
"""In-process compiler bridge.

Submits an emitted FrontendGraph to the native extension for verification,
canonical AIR lowering, and artifact compilation. These are compiler results,
not authoring APIs: authors write declarations and control flow, never a graph
dict.
"""

from __future__ import annotations

import json
from typing import Any, Optional

from ._native import compile_frontend_graph_artifact as _compile_frontend_graph_artifact
from ._native import lower_frontend_graph as _lower_frontend_graph
from ._native import verify_frontend_graph as _verify_frontend_graph

FRONTEND_GRAPH_VERSION = "apxm.frontend-graph.v1"
SOURCE_MAP_VERSION = "apxm.source-map.v1"


def verify_graph(graph: dict[str, Any]) -> Optional[str]:
    """Return ``None`` when the graph verifies, else a diagnostic string."""
    return _verify_frontend_graph(json.dumps(graph))


def lower_graph(graph: dict[str, Any]) -> dict[str, Any]:
    """Lower a FrontendGraph to canonical AIR through the native bridge."""
    return json.loads(_lower_frontend_graph(json.dumps(graph)))


def canonical_air_json(graph: dict[str, Any]) -> str:
    """The canonical AIR JSON string exactly as emitted by the native bridge."""
    return _lower_frontend_graph(json.dumps(graph))


def compile_artifact(graph: dict[str, Any]) -> dict[str, Any]:
    """Compile a FrontendGraph into one complete executable artifact."""
    return json.loads(_compile_frontend_graph_artifact(json.dumps(graph)))

File: crates/compiler/frontend/typescript/src/bridge.ts
// In-process compiler bridge over the native Node-API extension.
//
// Submits an emitted FrontendGraph for verification, canonical AIR lowering, and
// artifact compilation. These are compiler results, not authoring APIs: authors
// write declarations and control flow, never a graph object.

import { createRequire } from "node:module";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));

export const FRONTEND_GRAPH_VERSION = "apxm.frontend-graph.v1";
export const SOURCE_MAP_VERSION = "apxm.source-map.v1";

export type Json = Record<string, unknown>;

type NativeBridge = {
  compileFrontendGraphArtifact(json: string): string;
  lowerFrontendGraph(json: string): string;
  verifyFrontendGraph(json: string): string | null;
};

let loadedBridge: NativeBridge | undefined;

function nativeBridge(): NativeBridge {
  if (loadedBridge === undefined) {
    // The compiled bridge and the native extension are colocated; from source
    // (test runs) the extension is in the sibling build output.
    const candidates = [
      join(here, "_native.node"),
      join(here, "..", "dist", "_native.node"),
    ];
    let lastCause: unknown;
    for (const candidate of candidates) {
      try {
        loadedBridge = require(candidate) as NativeBridge;
        return loadedBridge;
      } catch (cause) {
        lastCause = cause;
      }
    }
    throw new Error(
      "The @apxm/frontend native compiler bridge is not installed for this platform",
      { cause: lastCause },
    );
  }
  return loadedBridge;
}

export function verifyGraph(graph: Json): string | null {
  return nativeBridge().verifyFrontendGraph(JSON.stringify(graph));
}

export function lowerGraph(graph: Json): Json {
  return JSON.parse(nativeBridge().lowerFrontendGraph(JSON.stringify(graph))) as Json;
}

export function canonicalAirJson(graph: Json): string {
  return nativeBridge().lowerFrontendGraph(JSON.stringify(graph));
}

export function compileArtifact(graph: Json): Json {
  return JSON.parse(
    nativeBridge().compileFrontendGraphArtifact(JSON.stringify(graph)),
  ) as Json;
}
